README.md
# node-bitcoin

node-bitcoin is a simple wrapper for the Bitcoin client's JSON-RPC API.

The API is equivalent to the API document [here](https://en.bitcoin.it/wiki/Original_Bitcoin_client/API_Calls_list).
The methods are exposed as lower camelcase methods on the `bitcoin.Client`
object, or you may call the API directly using the `cmd` method.

## Install

`npm install bitcoin`

## Setup

1. Traverse to `~/.bitcoin` or `~/Library/Application Support/Bitcoin` and add a
file called `bitcoin.conf` if it doesn't already exist.

2. Add these lines to the file:

    rpcuser=username

    rpcpassword=password

You will use these to login to the server.

3. Start your Bitcoin client with the `-server` argument or run `bitcoind`

4. You should now be able to communicate with Bitcoin JSON-RPC API using the
node-bitcoin library, try it out!

## Examples

### Create client
```js
var bitcoin = require('bitcoin');
var client = new bitcoin.Client('localhost', 8332, 'username', 'password');
```

### Create client with single object
```js
var client = new bitcoin.Client({
  host: 'localhost',
  port: 8332,
  user: 'username',
  pass: 'password'
});
```

### Get balance across all accounts with minimum confirmations of 6

```js
client.getBalance('*', 6, function(err, balance) {
  if (err) return console.log(err);
  console.log('Balance:', balance);
});
```
### Getting the balance directly using `cmd`

```js
client.cmd('getbalance', '*', 6, function(err, balance){
  if (err) return console.log(err);
  console.log('Balance:', balance);
});
```


Changelog.md
# node-bitcoin changelog

## v3.0.0 (2015/10/18)
* make public domain license explicit
* remove commands dropped in bitcoind v0.11
  * `getHashesPerSecond`
  * `getHashesPerSec`
* add missing commands for bitcoind v0.11
  * `generate`
  * `verifyTxOutProof`

## v2.4.0 (2015/07/16)
* don't lazy-load http/https modules
* add command for bitcoind v0.11.0: `getTxOutProof`

## v2.3.2 (2015/06/26)
* fix bug in test suite that was supposed to detect missing commands
* add missing commands
  * `prioritiseTransaction`
  * `importWallet`

## v2.3.1 (2015/06/24)
* add missing `getMempoolInfo` for bitcoind v0.10

## v2.3.0 (2015/02/04)
* drop node v0.8.x support
* update testnet-box
* update devDependencies
* add commands for bitcoind v0.10
  * `estimateFee`
  * `estimatePriority`
  * `getChainTips`
  * `importAddress`

## v2.2.0 (2014/08/29)
* add commands for bitcoind v0.9.x
  * `decodeScript`
  * `dumpWallet`
  * `getBestBlockHash`
  * `getBlockchainInfo`
  * `getNetTotals`
  * `getNetworkInfo`
  * `getNetworkHashPs`
  * `getRawChangeAddress`
  * `getUnconfirmedBalance`
  * `getWalletInfo`
  * `ping`
  * `verifyChain`

## v2.1.2 (2014/04/16)
* lazy load `http`/`https` module

## v2.1.1 (2014/03/25)
* change default request timeout from `5000`ms to `30000`ms

## v2.1.0 (2014/03/12)
* remove `deprecate` dependency
* add request timeout option (defaults to 5000ms)
* add 3rd parameter to callbacks: response headers

## v2.0.1 (2014/01/08)
* default `host` to 'localhost'; `port` to '8332'

## v2.0.0 (2013/10/14)
* remove deprecated commands
  * `getMemoryPool`
  * `getMemorypool`
* remove deprecated functionality
  * creating `bitcoin.Client` with more than one argument

## v1.7.0 (2013/05/05)
* add missing commands from bitcoind v0.7.0
  * `createMultiSig`
  * `getBlockTemplate`
  * `getTxOut`
  * `getTxOutSetInfo`
  * `listAddressGroupings`
  * `submitBlock`
* deprecate commands
  * `getMemoryPool`
  * `getMemorypool`

## v1.6.2 (2013/03/21)
* shrink package size via .npmignore

## v1.6.1 (2013/03/13)
* add node v0.10.x support (rejectUnauthorized defaults to true in 0.10.x)

## v1.6.0 (2013/03/08)
* drop node v0.6.x support
* change test runner from `vows` to `mocha`
* upgrade testnet-box
* add commands for bitcoind v0.8.0
  * `addNode`
  * `getAddedNodeInfo`
  * `listLockUnspent`
  * `lockUnspent`
* deprecate creating `bitcoin.Client` with more than one argument
* add SSL support

## v1.5.0 (2012/10/22)
* remove `getBlockNumber` test
* upgrade testnet-box
* add RPC call batching (multiple RPC calls within one HTTP request)

## v1.4.0 (2012/09/09)
* add commands for bitcoind v0.7.0
  * `createRawTransaction`
  * `decodeRawTransaction`
  * `getPeerInfo`
  * `getRawMemPool`
  * `getRawTransaction`
  * `listUnspent`
  * `sendRawTransaction`
  * `signRawTransaction`
* remove deprecated `getBlockNumber`

## v1.3.1 (2012/08/19)
Remove `underscore` dependency

## v1.3.0 (2012/07/03)
Change use of http.createClient() (deprecated in node v0.8.x) to http.request()

## v1.2.2 (2012/04/26)
Fix callback being called twice when a client and request error
occur on the same command call.

## v1.2.1 (2012/04/26)
* add missing `getBlock` command

## v1.2.0 (2012/04/25)
* submodule testnet-box for running tests
* err objects should all now be an instance of Error

## v1.1.6 (2012/04/11)
* add commands for bitcoind v0.6.0
  * `addMultiSigAddress` (only available in testnet)
  * `dumpPrivKey`
  * `getBlockHash`
  * `getMiningInfo`
  * `importPrivKey`


lib/jsonrpc.js
var http = require('http'),
    https = require('https');

var Client = function(opts) {
  this.opts = opts || {};
  this.http = this.opts.ssl ? https : http;
};

Client.prototype.call = function(method, params, callback, errback, path) {
  var time = Date.now();
  var requestJSON;

  if (Array.isArray(method)) {
    // multiple rpc batch call
    requestJSON = [];
    method.forEach(function(batchCall, i) {
      requestJSON.push({
        id: time + '-' + i,
        method: batchCall.method,
        params: batchCall.params
      });
    });
  } else {
    // single rpc call
    requestJSON = {
      id: time,
      method: method,
      params: params
    };
  }

  // First we encode the request into JSON
  var requestJSON = JSON.stringify(requestJSON);

  // prepare request options
  var requestOptions = {
    host: this.opts.host || 'localhost',
    port: this.opts.port || 8332,
    method: 'POST',
    path: path || '/',
    headers: {
      'Host': this.opts.host || 'localhost',
      'Content-Length': requestJSON.length
    },
    agent: false,
    rejectUnauthorized: this.opts.ssl && this.opts.sslStrict !== false
  };

  if (this.opts.ssl && this.opts.sslCa) {
    requestOptions.ca = this.opts.sslCa;
  }

  // use HTTP auth if user and password set
  if (this.opts.user && this.opts.pass) {
    requestOptions.auth = this.opts.user + ':' + this.opts.pass;
  }

  // Now we'll make a request to the server
  var cbCalled = false;
  var request = this.http.request(requestOptions);

  // start request timeout timer
  var reqTimeout = setTimeout(function() {
    if (cbCalled) return;
    cbCalled = true;
    request.abort();
    var err = new Error('ETIMEDOUT');
    err.code = 'ETIMEDOUT';
    errback(err);
  }, this.opts.timeout || 30000);

  // set additional timeout on socket in case of remote freeze after sending headers
  request.setTimeout(this.opts.timeout || 30000, function() {
    if (cbCalled) return;
    cbCalled = true;
    request.abort();
    var err = new Error('ESOCKETTIMEDOUT')
    err.code = 'ESOCKETTIMEDOUT'
    errback(err);
  });

  request.on('error', function(err) {
    if (cbCalled) return;
    cbCalled = true;
    clearTimeout(reqTimeout);
    errback(err);
  });

  request.on('response', function(response) {
    clearTimeout(reqTimeout);

    // We need to buffer the response chunks in a nonblocking way.
    var buffer = '';
    response.on('data', function(chunk) {
      buffer = buffer + chunk;
    });
    // When all the responses are finished, we decode the JSON and
    // depending on whether it's got a result or an error, we call
    // emitSuccess or emitError on the promise.
    response.on('end', function() {
      var err;

      if (cbCalled) return;
      cbCalled = true;

      try {
        var decoded = JSON.parse(buffer);
      } catch (e) {
        if (response.statusCode !== 200) {
          err = new Error('Invalid params, response status code: ' + response.statusCode);
          err.code = -32602;
          errback(err);
        } else {
          err = new Error('Problem parsing JSON response from server');
          err.code = -32603;
          errback(err);
        }
        return;
      }

      if (!Array.isArray(decoded)) {
        decoded = [decoded];
      }

      // iterate over each response, normally there will be just one
      // unless a batch rpc call response is being processed
      decoded.forEach(function(decodedResponse, i) {
        if (decodedResponse.hasOwnProperty('error') && decodedResponse.error != null) {
          if (errback) {
            err = new Error(decodedResponse.error.message || '');
            if (decodedResponse.error.code) {
              err.code = decodedResponse.error.code;
            }
            errback(err);
          }
        } else if (decodedResponse.hasOwnProperty('result')) {
          if (callback) {
            callback(decodedResponse.result, response.headers);
          }
        } else {
          if (errback) {
            err = new Error(decodedResponse.error.message || '');
            if (decodedResponse.error.code) {
              err.code = decodedResponse.error.code;
            }
            errback(err);
          }
        }
      });

    });
  });
  request.end(requestJSON);
};

module.exports.Client = Client;
